Use _as_mpf_val methods when creating new Floats

cbm755 opened the pull request:
This adds support for `Float(S.Pi)` (which previously gave an error).

Having done that, I noticed the same mechanisgm could de-dupes the code a bit.  I don't think this has any performance implications; same code is being called, just in the `_as_mpf_val` method now.

cbm755 committed:
Float(Expr) should work for anything with a ._as_mpf_val

There is a last-ditch effort to try `mpmath.mpf` but no effort was
made for certain SymPy objects such as "S.Pi" or "S.Catalan", even
though they can easily be made into mpf.

cbm755 committed:
Float: let the catch-all handle Rational and Float

De-dupe the code a bit.  Rational and Float objects know how to
convert themselves to mpf via the `_as_mpf_val()` method, so the
Float constructor does not need to repeat that logic.

cbm755 committed:
Integer: use same rounding as other objects for mpf

Fix failing tests.  Commit 92544afb631f514d4d3c0af120da2fe0e5f5894c
means that `Integer._as_mpf_val` is used to convert large Integers
to low-precision Floats, whereas previously Integers were treated
the same as Rationals in the Float constructor.  This fixes the
Integer method to behave the same as the Rational (and other)
`_as_mpf_val`.

jksuom commented:
This looks good but the geometry failure is puzzling. The code evaluates `sqrt(10 - 7*sqrt(2))` using the low-level method `_eval_evalf`:
```
>>> e = sqrt(10 - 7*sqrt(2))
>>> e.evalf(2)
0.32
>>> e._eval_evalf(2)
0.e+0*I
>>> srepr(_)
"Mul(Float('2.0', precision=2), I)"
```
Master behaves differently but it also gives an unexpected result:
```
>>> e = sqrt(10 - 7*sqrt(2))
>>> e.evalf(2)
0.32
>>> e._eval_evalf(2)
0
>>> srepr(_)
'Integer(0)'
```

cbm755 commented:
Thanks for helping narrow that down!

Something to keep in mind is that `evalf()` takes decimal digits, but `_eval_evalf` takes binary precision, so those `2`s are not the same thing.

I'll try to dig into this a bit.  So far, I'm not very convinced by these various 2-binary-digit float checks deep down in routines like `.is_positive`...

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -1946,7 +1943,7 @@ class Integer(Rational):
 
     def _as_mpf_val(self, prec):
-        return mlib.from_int(self.p, prec)
+        return mlib.from_int(self.p, prec, rnd)
```
this change makes the polygon test fail.  If that test is so fragile that changing the rounding mode causes it to fail, then I'm inclined to blame the test ;-)

A nice kind of fuzzing in numerical analysis is to change the IEEE rounding mode and rerun the code.  Might be nice to try that in SymPy (mpmath really) and see how many of our tests break!

jksuom replied:
It seems that the test was changed. The [original test](https://github.com/sympy/sympy/commit/c5f0b00f#diff-c5cd274236c00ee31386e2953a14a438L1651) looked more robust to me. But maybe something made that fail too.

asmeurer commented:
> I'll try to dig into this a bit. So far, I'm not very convinced by these various 2-binary-digit float checks deep down in routines like .is_positive...

There are some known issues (across several issue in the tracker, but https://github.com/sympy/sympy/issues/12534 is a good place to start). From what I could tell when working with it is that the code is correct, but if you get something like 0.e-1 it means it's indeterminant (I forget exactly what it looks like, but it ought to made a lot more clear).

cbm755 committed:
Segment: try a bit harder to simplify for contains Point check

Move the simplify outside of the Eq which IIRC can sometimes evaluate
further than Eq on the simplified inputs.  At any rate, this seems to
fix a failing test and the Eq argument will still be simplified.

jksuom commented:
+1 for this minimal fix.

I suspect that the critical test in `contains` has to be redesigned at some point. It seems to me that testing the dot products would have been more robust. I wonder why that was replaced by the "triangle equality" test. cc @siefkenj

siefkenj commented:
I don't remember why I changed it to a triangle inequality test. Usually I changed things because they were specific to 2D or they failed with symbols or some such, but I don't the reason for the change now. We could change it back and see if any of the tests fail.

jksuom commented:
Thanks. This looks good to me now. That test can be handled in another PR.

cbm755 commented:
> We could change it back and see if any of the tests fail.

Shall we continue that discussion in #13314?
